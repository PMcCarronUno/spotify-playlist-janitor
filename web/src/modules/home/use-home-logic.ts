import { useCallback, useEffect, useState } from "react";

import { useDataApi } from "api/data-api";
import { useSpotifyApi } from "api/spotify-api";

import { Playlist } from "./home-types";
import { useModal } from "shared/components/modal";

export const useHomeLogic = () => {
  const { getDatabasePlaylists, addDatabasePlaylist } = useDataApi();
  const { getSpotifyPlaylists } = useSpotifyApi();

  const [loading, setLoading] = useState(false);
  const [modalSaving, setModalSaving] = useState(false);
  const [showSpotifyAuthModal, setShowSpotifyAuthModal] = useState(false);
  const [monitoredPlaylists, setMonitoredPlaylists] = useState<
    Playlist[] | undefined
  >(undefined);
  const [unmonitoredPlaylists, setUnmonitoredPlaylists] = useState<
    Playlist[] | undefined
  >(undefined);
  const [modalError, setModalError] = useState<string | undefined>(undefined);

  const { isOpen: modalOpen, onOpen: onModalOpen, onClose } = useModal();

  const onModalClose = useCallback(() => {
    setModalError(undefined);
    onClose();
  }, [setModalError, onClose]);

  const getPlaylistData = useCallback(async () => {
    try {
      setLoading(true);
      const databasePlaylistsResponse = await getDatabasePlaylists();
      const spotifyPlaylistsResponse = await getSpotifyPlaylists();

      setMonitoredPlaylists(
        spotifyPlaylistsResponse.data
          .filter((spotifyPlaylist) =>
            databasePlaylistsResponse.data.some(
              (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
            )
          )
          .map((monitoredPlaylist) => ({
            id: monitoredPlaylist.id,
            href: monitoredPlaylist.href,
            name: monitoredPlaylist.name,
            image: monitoredPlaylist.images[0],
          }))
      );

      setUnmonitoredPlaylists(
        spotifyPlaylistsResponse.data
          .filter(
            (spotifyPlaylist) =>
              !databasePlaylistsResponse.data.some(
                (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
              )
          )
          .map((unmonitoredPlaylist) => ({
            id: unmonitoredPlaylist.id,
            href: unmonitoredPlaylist.href,
            name: unmonitoredPlaylist.name,
            image: unmonitoredPlaylist.images[0],
          }))
      );
      setLoading(false);
    } catch (e: any) {
      if (
        e?.response?.status === 500 &&
        e?.response?.data?.message ===
          "Application has not been logged into your Spotify account."
      ) {
        setShowSpotifyAuthModal(true);
      }
      setLoading(false);
    }
  }, [
    setLoading,
    getDatabasePlaylists,
    getSpotifyPlaylists,
    setMonitoredPlaylists,
    setUnmonitoredPlaylists,
  ]);

  useEffect(() => {
    getPlaylistData();
  }, [getPlaylistData]);

  const onSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      setModalError(undefined);
      const playlistId = (event.currentTarget[1] as HTMLInputElement).value;
      const skipThreshold = (event.currentTarget[2] as HTMLInputElement).value;
      const ignoreInitialSkips = (event.currentTarget[3] as HTMLInputElement)
        .checked;
      const autoDeleteTracksAfter = (event.currentTarget[4] as HTMLInputElement)
        .value;

      if (!playlistId) {
        setModalError("Please select a playlist");
        return;
      }
      if (!!autoDeleteTracksAfter && parseInt(autoDeleteTracksAfter) === 0) {
        setModalError("Auto-delete must be empty or greater than 0");
        return;
      }

      try {
        setModalSaving(true);
        
        const request = {
          id: playlistId,
          skipThreshold: !!skipThreshold ? parseInt(skipThreshold) : undefined,
          ignoreInitialSkips: ignoreInitialSkips,
          autoCleanupLimit: !!autoDeleteTracksAfter
            ? parseInt(autoDeleteTracksAfter)
            : undefined,
        };

        await addDatabasePlaylist(request);
        setModalSaving(false);
        onModalClose();
        await getPlaylistData();
      } catch (e: any) {
        setModalError(e.response?.data?.message || "Unknown error");
        setModalSaving(false);
      }
    },
    [addDatabasePlaylist, getPlaylistData, onModalClose]
  );

  const onPlaylistChange = useCallback(
    async (newValue: { label: string; value: string }) => {
      if (newValue?.value) {
        setModalError(undefined);
      }
    },
    []
  );

  return {
    monitoredPlaylists,
    unmonitoredPlaylists,
    loading,
    modalOpen,
    onModalOpen,
    onModalClose,
    onSubmit,
    onPlaylistChange,
    modalError,
    modalSaving,
    showSpotifyAuthModal,
  } as const;
};
