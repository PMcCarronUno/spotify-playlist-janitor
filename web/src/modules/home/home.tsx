import styled from "styled-components";
import { AiOutlinePlusCircle } from "react-icons/ai";

import { BLACK, GREEN, RED } from "shared/constants";
import { SubTitle, Text } from "shared/components/typography";
import { Modal } from "shared/components/modal";

import { useHomeLogic } from "./use-home-logic";
import { AddPlaylistModalView } from "./modal/add-playlist-modal";

export const Home = () => {
  const {
    monitoredPlaylists,
    unmonitoredPlaylists,
    loading,
    modalOpen,
    onModalOpen,
    onModalClose,
    onSubmit,
    onPlaylistChange,
    modalError,
    modalSaving,
    showSpotifyAuthModal,
  } = useHomeLogic();

  return (
    <PageContainer>
      {monitoredPlaylists && (
        <>
          {monitoredPlaylists.map((monitoredPlaylist, index) => (
            <div
              key={monitoredPlaylist.id}
              className="item"
              onClick={() =>
                alert(
                  `Naviagte to ${monitoredPlaylist.name} at route ${monitoredPlaylist.id}`
                )
              }
              id={`playlist-item-${index}`}
              data-testid={`playlist-item-${index}`}
            >
              {monitoredPlaylist.image && (
                <img
                  alt={monitoredPlaylist.name}
                  width={130}
                  height={130}
                  src={monitoredPlaylist.image.url}
                />
              )}

              <SubTitle>{monitoredPlaylist.name}</SubTitle>
            </div>
          ))}
          <div
            className="item new"
            onClick={onModalOpen}
            id={`add-playlist-item`}
            data-testid={`add-playlist-item`}
          >
            <AiOutlinePlusCircle />
          </div>
        </>
      )}
      <Modal
        {...{
          isOpen: modalOpen,
          onClose: onModalClose,
          label: "Select Playlist Modal",
        }}
      >
        <AddPlaylistModalView
          {...{
            onSubmit,
            unmonitoredPlaylists,
            onPlaylistChange,
            modalSaving,
            modalError,
            onModalClose,
          }}
        />
      </Modal>
      <Modal
        {...{
          isOpen: showSpotifyAuthModal,
          onClose: () => {},
          label: "Spotify Auth Modal",
        }}
      >
        <ModalContainer>
          <SubTitle style={{ marginBottom: 8 }}>
            Spotify Authentication Error
          </SubTitle>
          <Text>
            The application has not been authenticated with your Spotify
            account. Please follow{" "}
            <a
              href={process.env.REACT_APP_API_URL}
              target="_blank"
              rel="noreferrer"
            >
              this
            </a>{" "}
            link and sign into Spotify, and <a href=".">reload</a> this page.
          </Text>
        </ModalContainer>
      </Modal>
    </PageContainer>
  );
};

const PageContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  align-content: flex-start;

  color: white;
  width: 100%;
  max-width: 1825px;
  paddingtop: 18px;

  .item {
    flex: 1 1 30%; /*grow | shrink | basis */
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    margin: 8px;
    height: 160px;
    max-width: 33%;
    background-color: ${BLACK};

    border-radius: 15px;

    cursor: pointer;

    &:hover {
      transform: scale(1.01);
      box-shadow: 0 0 32px rgba(117, 117, 117, 0.2);
    }
  }

  .new {
    background-color: ${GREEN};

    svg {
      height: 6em;
      width: 6em;
    }
  }
`;

const ModalContainer = styled.div`
  width: 360px;

  .rows {
    margin-bottom: 8px;
  }

  .row {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-bottom: 8px;
  }

  .number-input {
    width: 55px !important;
  }

  .toggle-input {
    margin-right: 16px;
  }

  .auto-delete-input {
    margin-right: -10px;
  }

  .auto-delete-post-text {
    margin-right: 10px;
  }

  .error {
    color: ${RED};
  }
`;

Home.displayName = "Home";
