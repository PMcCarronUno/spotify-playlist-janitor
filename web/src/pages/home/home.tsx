import styled from "styled-components";
import { HEADER_HEIGHT, HEADER_PADDING } from "shared/constants";
import { Home as HomeInitialise } from "modules/home";

export const Home = () => {
  return (
    <Content>
      <HomeInitialise />
    </Content>
  );
};

const Content = styled.div`
  display: flex;
  justify-content: center;
  height: calc(
    100vh - ${HEADER_HEIGHT} - ${HEADER_PADDING} - ${HEADER_PADDING}
  );
`;

Home.displayName = "Home";
