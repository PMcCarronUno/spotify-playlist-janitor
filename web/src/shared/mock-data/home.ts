import { Playlist } from "modules/home/home-types";
import { databasePlaylists, spotifyPlaylists } from "shared/mock-data/api";

export const monitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter((spotifyPlaylist) =>
    databasePlaylists.some(
      (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
    )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));

export const unmonitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter(
    (spotifyPlaylist) =>
      !databasePlaylists.some(
        (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
      )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));
