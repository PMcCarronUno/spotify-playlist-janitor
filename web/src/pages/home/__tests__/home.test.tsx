import React from 'react';
import { render } from '@testing-library/react';

import { Home } from '../home';

describe('<Home />', () => {
  let container: HTMLElement;

  beforeEach(() => {
    ({ container } = render(<Home />));
  });

  it('should render a Home component', () => {
    expect(container.firstChild).toMatchSnapshot();
  });
});