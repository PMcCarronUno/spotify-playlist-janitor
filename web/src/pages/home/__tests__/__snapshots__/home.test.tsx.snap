// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render a Home component 1`] = `
<div
  class="sc-beyTiQ elccEX"
>
  API URL: 
  https://localhost:5001
</div>
`;
